File: backend/app/db_models/prediction.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class PredictionJob(Base):
    __tablename__ = "prediction_jobs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True, nullable=False)
    model_id: Mapped[int] = mapped_column(ForeignKey("models.id"), index=True, nullable=False)
    dataset_id: Mapped[int | None] = mapped_column(ForeignKey("datasets.id"), nullable=True)
    job_type: Mapped[str] = mapped_column(String(32), nullable=False)
    input_file_path: Mapped[str | None] = mapped_column(String(512), nullable=True)
    output_file_path: Mapped[str | None] = mapped_column(String(512), nullable=True)
    status: Mapped[str] = mapped_column(String(32), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)


class PredictionResult(Base):
    __tablename__ = "prediction_results"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    job_id: Mapped[int] = mapped_column(ForeignKey("prediction_jobs.id"), index=True, nullable=False)
    sample_index: Mapped[int] = mapped_column(Integer, nullable=False)
    input_json: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    prediction_json: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)

File: frontend/src/types/prediction.ts
export type PredictionLabelResult = {
  label: string;
  probability: number | null;
};

export type SinglePredictionPayload = {
  model_id: number;
  input_data: Record<string, unknown>;
};

export type SinglePredictionResponse = {
  job_id: number;
  task: string;
  result: Record<string, PredictionLabelResult>;
  disclaimer: string;
};

export type BatchPredictionResponse = {
  job_id: number;
  rows: Record<string, unknown>[];
  disclaimer: string;
};

export type PredictionJob = {
  id: number;
  user_id: number;
  model_id: number;
  dataset_id: number | null;
  job_type: string;
  input_file_path: string | null;
  output_file_path: string | null;
  status: string;
  created_at: string;
  finished_at: string | null;
  error_message: string | null;
};
